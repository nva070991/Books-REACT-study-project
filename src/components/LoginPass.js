import {
  useContext, useState, React, useRef,
} from 'react'
import UserContext from '../contexts/Login'

function LoginPass() {
  const inputLog = useRef(null)
  const inpuPass = useRef(null)
  const context = useContext(UserContext)

  const [log, SetLog] = useState('')
  const [pass, SetPass] = useState('')
  const [logIsHidden, setlogIsHidde] = useState(false)
  const [pasIsHidden, setpasIsHidden] = useState(false)
  const [messageLog, SetMessageLog] = useState('')
  const [messagePas, SetMessagePas] = useState('')

  const Submit = () => {
    if (log === '' || (!(/\S+@\S+\.\S+/.test(log)))) {
      inputLog.current.focus()
    } else if (pass === '' || pass.length < 6) {
      inpuPass.current.focus()
    } else {
      context.SetLogin(log)
      context.SetPassword(pass)
      localStorage.user = log
      localStorage.password = pass
    }
  }

  const Parse = (id) => {
    if (id === 'login') {
      const { value } = inputLog.current

      SetLog(value)
    } else if (id === 'password') {
      const { value } = inpuPass.current

      SetPass(value)
    }
  }

  const Hide = (id) => {
    if (id === 'login') {
      if (log === '') {
        SetMessageLog('обязательно поле')
        setlogIsHidde(true)
      } else if (!(/\S+@\S+\.\S+/.test(log))) {
        SetMessageLog('введите e-mail')
        setlogIsHidde(true)
      } else {
        SetMessageLog('')
        setlogIsHidde(false)
      }
    } else if (id === 'password') {
      if (pass === '') {
        SetMessagePas('обязательно поле')

        setpasIsHidden(true)
      } else if (pass.length < 6) {
        SetMessagePas('минимальная длина пароля 6 символов')
        setpasIsHidden(true)
      } else {
        SetMessagePas('')
        setpasIsHidden(false)
      }
    }
  }

  const useInputRequired = (e) => {
    if (e.type === 'change') {
      Parse(e.target.id)
    }
    if (e.type === 'blur') {
      Hide(e.target.id)
    }
    if (e.type === 'submit') {
      e.preventDefault()

      Hide('login')
      Parse('login')
      Hide('password')
      Parse('password')

      Submit()
    }
  }

  return (
    <div className="DivPadding">

      <form onSubmit={useInputRequired}>
        <div>
          <div style={{ width: '250px', display: 'flex', justifyContent: 'space-between' }}>
            Login
            <input data-testid="login" ref={inputLog} type="text" value={log} id="login" onBlur={useInputRequired} onChange={useInputRequired} />

          </div>
          <div>
            {!logIsHidden
              ? ('')
              : (
                <div
                  data-testid="loghint"
                  style={{
                    width: '250px', display: 'flex', flexDirection: 'row-reverse', color: 'red', fontSize: '12px',
                  }}
                >
                  {messageLog}
                </div>
              )}

          </div>
          <div style={{ width: '250px', display: 'flex', justifyContent: 'space-between' }}>
            Password
            <input data-testid="password" ref={inpuPass} type="text" value={pass} id="password" onBlur={useInputRequired} onChange={useInputRequired} />

          </div>
          <div>
            {!pasIsHidden
              ? ('')
              : (
                <div
                  data-testid="pashint"
                  style={{
                    width: '250px', display: 'flex', flexDirection: 'row-reverse', color: 'red', fontSize: '12px',
                  }}
                >
                  {messagePas}
                </div>
              )}

          </div>
        </div>
        <button data-testid="button" type="submit">Войти</button>
      </form>
    </div>
  )
}

export default LoginPass
