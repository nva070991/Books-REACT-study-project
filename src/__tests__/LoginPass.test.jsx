import { render, screen, fireEvent } from '@testing-library/react'
import { Routes, Route, BrowserRouter } from 'react-router-dom'
import React from 'react'
import UserContext from '../contexts/Login'

import LoginPass from '../components/LoginPass'

let login = 1
const SetLogin = (l) => { login = l }
let Password = 1
const SetPassword = (l) => { Password = l }

describe('LoginPass component', () => {
  it('should render LoginPass component', () => {
    const view = render(
      <BrowserRouter>
        <UserContext.Provider value={{
          login, SetLogin, Password, SetPassword,
        }}
        >

          <Routes>
            <Route path="/" element={<LoginPass />} />

          </Routes>
        </UserContext.Provider>
      </BrowserRouter>,
    )
    expect(view).toMatchSnapshot()
  })
  it('should work tips', () => {
    render(
      <BrowserRouter>
        <UserContext.Provider value={{
          login, SetLogin, Password, SetPassword,
        }}
        >

          <Routes>
            <Route path="/" element={<LoginPass />} />

          </Routes>
        </UserContext.Provider>
      </BrowserRouter>,
    )

    const button = screen.getByTestId('button')

    fireEvent.click(button)

    expect(screen.getByTestId('pashint')).toHaveTextContent('обязательно поле')
    expect(screen.getByTestId('loghint')).toHaveTextContent('обязательно поле')
  })
})
