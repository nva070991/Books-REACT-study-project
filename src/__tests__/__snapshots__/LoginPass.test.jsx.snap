// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPass component should render LoginPass component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="DivPadding"
      >
        <form>
          <div>
            <div
              style="width: 250px; display: flex; justify-content: space-between;"
            >
              Login
              <input
                data-testid="login"
                id="login"
                type="text"
                value=""
              />
            </div>
            <div />
            <div
              style="width: 250px; display: flex; justify-content: space-between;"
            >
              Password
              <input
                data-testid="password"
                id="password"
                type="text"
                value=""
              />
            </div>
            <div />
          </div>
          <button
            data-testid="button"
            type="submit"
          >
            Войти
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="DivPadding"
    >
      <form>
        <div>
          <div
            style="width: 250px; display: flex; justify-content: space-between;"
          >
            Login
            <input
              data-testid="login"
              id="login"
              type="text"
              value=""
            />
          </div>
          <div />
          <div
            style="width: 250px; display: flex; justify-content: space-between;"
          >
            Password
            <input
              data-testid="password"
              id="password"
              type="text"
              value=""
            />
          </div>
          <div />
        </div>
        <button
          data-testid="button"
          type="submit"
        >
          Войти
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
